daemon/cmd/state.go

```diff
@@ -90,6 +91,39 @@ func (d *Daemon) validateEndpoint(ep *endpoint.Endpoint) (valid bool, err error)
 	return true, nil
 }
 
+// fetchOldEndpoints reads the list of existing endpoints previously managed by Cilium when it was
+// last run and associated it with container workloads. This function performs the first step in
+// restoring the endpoint structure.  It needs to be followed by a call to restoreOldEndpoints()
+// once k8s has been initialized and regenerateRestoredEndpoints() once the endpoint builder is
+// ready.
+func (d *Daemon) fetchOldEndpoints(dir string) (*endpointRestoreState, error) {
+	state := &endpointRestoreState{
+		possible: nil,
+		restored: []*endpoint.Endpoint{},
+		toClean:  []*endpoint.Endpoint{},
+	}
+
+	if !option.Config.RestoreState {
+		log.Info("Endpoint restore is disabled, skipping restore step")
+		return state, nil
+	}
+
+	log.Info("Reading old endpoints...")
+
+	dirFiles, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return state, err
+	}
+	eptsID := endpoint.FilterEPDir(dirFiles)
+
+	state.possible = endpoint.ReadEPsFromDirNames(d.ctx, d, dir, eptsID)
```

Comment:
Might be worth to change `possible` to be `[]*endpoint.Endpoint` and convert the return value of `endpoint.ReadEPsFromDirNames` to it in order to avoid the random iteration order issue pointed out above, e.g.:

```suggestion
	possible = endpoint.ReadEPsFromDirNames(d.ctx, d, dir, eptsID)
	state.possible = make([]*endpoint.Endpoint, 0, len(possible)
	for _, p := range possible {
		state.possible = append(state.possible, p)
	}
	sort.Slice(state.possible, func(i, j int) bool {
		return state.possible[i].ID < state.possible[j].ID
	}
```

Reply:
As said above, the iteration order should make no difference.